File: src/Python/examples/finance_steps.py
from __future__ import annotations
from dataclasses import dataclass

@dataclass(frozen=True)
class Tick:
    symbol: str
    price: float

@dataclass(frozen=True)
class Features:
    r1: float
    vol: float

@dataclass(frozen=True)
class Score:
    value: float

class OrderResponse: pass

@dataclass(frozen=True)
class Accept(OrderResponse):
    symbol: str
    qty: int
    price: float

@dataclass(frozen=True)
class Reject(OrderResponse):
    reason: str

class FinanceSteps:
    @staticmethod
    def compute_features(t: Tick) -> Features:
        r1 = 0.0
        vol = abs(t.price) * 0.01
        return Features(r1, vol)

    @staticmethod
    def score(f: Features) -> Score:
        return Score(max(0.0, 1.0 - f.vol))

    @staticmethod
    def decide(s: Score) -> OrderResponse:
        return Accept("AAPL", 10, 101.25) if s.value >= 0.5 else Reject("LowScore")

File: src/typescript/examples/finance_steps.ts
export interface Tick {
  symbol: string;
  price: number;
}

export interface Features {
  r1: number;
  vol: number;
  symbol: string;
  price: number;
}

export interface Score {
  value: number;
}

export type OrderResponse = Accept | Reject;

export interface Accept {
  type: "accept";
  symbol: string;
  qty: number;
  price: number;
}

export interface Reject {
  type: "reject";
  reason: string;
}

export class FinanceSteps {
  static computeFeatures(tickValue: Tick): Features {
    const volatility = Math.abs(tickValue.price) * 0.01;
    return { r1: 0.0, vol: volatility, symbol: tickValue.symbol, price: tickValue.price };
  }

  static score(featuresValue: Features): Score {
    const scoreValue = Math.max(0.0, 1.0 - featuresValue.vol);
    return { value: scoreValue };
  }

  static decide(scoreValue: Score): OrderResponse {
    if (scoreValue.value >= 0.5) {
      return { type: "accept", symbol: "AAPL", qty: 10, price: 101.25 };
    } else {
      return { type: "reject", reason: "LowScore" };
    }
  }
}
